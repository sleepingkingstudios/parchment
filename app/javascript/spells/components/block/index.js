/* eslint-disable-next-line import/prefer-default-export */
export { default as Block } from './block';
