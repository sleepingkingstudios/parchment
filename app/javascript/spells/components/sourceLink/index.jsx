import React from 'react';
import PropTypes from 'prop-types';
import { Link } from 'react-router-dom';
import pluralise from 'pluralise';

import { exists } from '../../../utils/object';
import { underscore } from '../../../utils/string';

const emptySource = () => (
  <span className="text-muted">Homebrew</span>
);

const sourceLink = ({ id, sourceType }) => {
  const sourcePath = pluralise(2, underscore(sourceType));

  return `/${sourcePath}/${id}`;
};

const SpellSourceLink = ({ source, sourceType }) => {
  if (!exists(source) || !exists(sourceType)) { return emptySource(); }

  const { id, name } = source;

  return (
    <Link to={sourceLink({ id, sourceType })}>{ name }</Link>
  );
};

SpellSourceLink.defaultProps = {
  source: null,
  sourceType: null,
};

SpellSourceLink.propTypes = {
  source: PropTypes.shape({
    id: PropTypes.string,
    name: PropTypes.string,
  }),
  sourceType: PropTypes.string,
};

export default SpellSourceLink;
