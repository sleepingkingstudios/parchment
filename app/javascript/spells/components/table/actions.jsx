import PropTypes from 'prop-types';

import ResponsiveActions from '../../../components/responsive-actions';
import deleteEndpoint from '../../store/deleteSpell';
import { injectProps } from '../../../utils/react';

const SpellsTableActions = injectProps(
  ResponsiveActions,
  {
    deleteEndpoint,
    resourceName: 'spell',
  },
);

SpellsTableActions.defaultProps = {
  onDelete: () => {},
};

SpellsTableActions.propTypes = {
  id: PropTypes.string.isRequired,
  onDelete: PropTypes.func,
};

export default SpellsTableActions;
