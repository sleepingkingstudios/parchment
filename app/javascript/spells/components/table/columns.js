import ResponsiveActions from 'components/responsive-actions';
import { injectProps } from 'utils/react';
import { truncate } from 'utils/string';
import ResponsiveSchool from './responsive-school';

const spellDescription = (spell) => {
  if (spell.shortDescription) {
    return truncate(spell.shortDescription, 80);
  }

  return truncate(spell.description, 80);
};

const generateColumns = ({ actions, baseUrl, useDestroyRequest }) => ([
  {
    label: 'Name',
    prop: 'name',
    width: 6,
  },
  {
    label: 'School',
    prop: 'school',
    width: 3,
    value: ResponsiveSchool,
  },
  {
    label: false,
    prop: 'actions',
    width: 3,
    value: injectProps(
      ResponsiveActions,
      {
        actions,
        baseUrl,
        resourceName: 'spell',
        useDestroyRequest,
      },
    ),
  },
  {
    header: false,
    label: false,
    prop: 'description',
    width: 12,
    value: spellDescription,
  },
]);

export default generateColumns;
