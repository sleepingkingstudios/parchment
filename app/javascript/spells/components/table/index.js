export { default as columns } from './columns';

export { default as Table } from './table';
