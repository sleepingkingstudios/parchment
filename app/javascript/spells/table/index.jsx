import React from 'react';
import PropTypes from 'prop-types';

import spellType from '../spell';
import TableBody from './body';
import TableHeader from './header';

const SpellsTable = (props) => {
  const { spells, spellsRequestStatus } = props;

  return (
    <table>
      <TableHeader />

      <TableBody {...{ spells, spellsRequestStatus }} />
    </table>
  );
};

SpellsTable.defaultProps = {};

SpellsTable.propTypes = {
  spells: PropTypes.arrayOf(spellType).isRequired,
  spellsRequestStatus: PropTypes.string.isRequired,
};

export default SpellsTable;
