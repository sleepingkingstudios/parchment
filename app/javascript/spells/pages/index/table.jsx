import React from 'react';

import { LoaderSwitch } from '../../../components/loader';
import { SpellsTable } from '../../components/table';
import { useSpells } from '../../store/indexFindSpells';

const renderFailure = () => (
  <p>Unable to load spells data from the server.</p>
);
const renderPending = () => (
  <p>Loading spells data from the server...</p>
);

const IndexSpellsTable = () => {
  const { spells, status } = useSpells();

  return (
    <LoaderSwitch
      renderFailure={renderFailure}
      renderInitialized={renderPending}
      renderPending={renderPending}
      renderSuccess={() => (<SpellsTable spells={spells} />)}
      status={status}
    />
  );
};

IndexSpellsTable.defaultProps = {};

IndexSpellsTable.propTypes = {};

export default IndexSpellsTable;
