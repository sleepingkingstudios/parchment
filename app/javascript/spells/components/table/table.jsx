import React from 'react';
import PropTypes from 'prop-types';

import Table from '../../../components/table';
import columns from './columns';
import { spellListType } from '../../entities';

const emptyMessage = 'There are no spells matching the criteria.';

const SpellsTable = ({ spells, onDelete }) => (
  <Table columns={columns} data={spells} message={emptyMessage} cellProps={{ onDelete }} />
);

SpellsTable.defaultProps = {
  onDelete: () => {},
};

SpellsTable.propTypes = {
  onDelete: PropTypes.func,
  spells: spellListType.isRequired,
};

export default SpellsTable;
