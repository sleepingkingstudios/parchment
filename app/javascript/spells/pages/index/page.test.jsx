import React from 'react';
import { shallow } from 'enzyme';

import IndexSpellsPage from './page';
import DynamicTable from '../../../components/dynamic-table';
import { columns } from '../../components/table';
import endpoint from '../../store/indexFindSpells';

describe('IndexSpellsPage', () => {
  const defaultProps = {};

  it('should render the index page', () => {
    const rendered = shallow(<IndexSpellsPage {...defaultProps} />);

    expect(rendered).toHaveDisplayName('IndexPage');
    expect(rendered).toHaveProp({ columns });
    expect(rendered).toHaveProp({ endpoint });
    expect(rendered).toHaveProp({ resourceName: 'Spell' });
    expect(rendered).toHaveProp({ Table: DynamicTable });
  });
});
