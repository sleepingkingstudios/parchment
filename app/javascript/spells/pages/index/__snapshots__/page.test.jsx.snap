// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SpellsPage /> should match the snapshot 1`] = `
<Page
  breadcrumbs={
    Array [
      Object {
        "label": "Home",
        "url": "/",
      },
      Object {
        "active": true,
        "label": "Spells",
        "url": "/spells",
      },
    ]
  }
  className="page-spells"
  layout={null}
  subtitle="5e Campaign Companion"
  title="Parchment"
>
  <HeadingWithButtons
    buttons={
      Array [
        Object {
          "label": "Create Spell",
          "outline": true,
          "url": "/spells/create",
        },
      ]
    }
  >
    Spells
  </HeadingWithButtons>
  <IndexSpellsTable />
</Page>
`;
